Various tutorial fixes from feedback (#2674)
unicode-org/icu4x: docs/tutorials/intro.md

To use `ICU4X` in the Rust ecosystem one can either add dependencies on selected components, or add a dependency on the meta-crate `icu` which brings a reasonable selection of components in the most user-friendly configuration of features.

In this tutorial we are going to build up to writing an app that uses the `icu::datetime` component to format a date and time in a Japanese locale, covering various topics in the process.

# 1. Requirements

For this tutorial we assume the user has basic Rust knowledge. If acquiring it is necessary, the [Rust Book](https://doc.rust-lang.org/book/) provides an excellent introduction.
We also assume that the user is familiar with a terminal and have `rust` and `cargo` installed.

To verify that, open a terminal and check that the results are similar to:

```
$ cargo --version
cargo 1.51.0 (43b129a20 2021-03-16)
```

In this tutorial we are going to use a directory relative to the user's home directory `~/projects/icu_tutorial/`. 

# 2. Creating MyApp

To initialize a binary application, navigate to the root directory of our project and initialize a new binary app called `myapp`:

```
cd ~/projects/icu_tutorial
cargo new --bin myapp
cd myapp
```

> The result is a new directory `~/projects/icu_tutorial/myapp` with a file `./src/main.rs` which is the main file for our application. For most of this tutorial we'll be working from the `myapp` folder.

Reviewer: Nit: I've never been comfortable with the `icu` name since it conflicts with regular ICU, but I'm living with it on crates.io since it is a Rust ecosystem thing and we reached consensus long ago to go with that name. However, when creating any directories or other identifiers, I would really prefer if we would stick with `icu4x` as our primary identifier.

To use `ICU4X` in the Rust ecosystem one can either add dependencies on selected components, or add a dependency on the meta-crate `icu` which brings a reasonable selection of components in the most user-friendly configuration of features.

In this tutorial we are going to build up to writing an app that uses the `icu::datetime` component to format a date and time in a Japanese locale, covering various topics in the process.

# 1. Requirements

For this tutorial we assume the user has basic Rust knowledge. If acquiring it is necessary, the [Rust Book](https://doc.rust-lang.org/book/) provides an excellent introduction.
We also assume that the user is familiar with a terminal and have `git`, `rust` and `cargo` installed.

To verify that, open a terminal and check that the results are similar to:

```

> user@host:~/projects/icu_tutorial$ git --version

Reviewer 2: We're not using `git` in this tutorial.

git version 2.31.1

> user@host:~/projects/icu_tutorial$ cargo --version

Reviewer 2: ```suggestion
$ cargo --version
```

cargo 1.51.0 (43b129a20 2021-03-16)
```

> In this tutorial we are going to use a directory relative to the user's home directory `~/projects/icu_tutorial/`. The `~` in the path indicates the relative location of the user home directory.

Reviewer 2: If we're assuming `git` and `cargo` knowledge, we can also assume that the user knows what `~` means. But let's not use absolute paths though:

```
# 2. Creating MyApp

In your working directory, use `cargo` to create a binary project:

\```
cargo new --bin myapp
cd myapp
\```
```

# 2. Creating MyApp

To initialize a binary application, navigate to the root directory of our project and initialize a new binary app called `myapp`:

```
cd ~/projects/icu_tutorial
cargo new --bin myapp
```

> The result is a new directory `~/projects/icu_tutorial/myapp` with a file `./src/main.rs` which is the main file for our application.

Reviewer 2: If we're assuming `cargo` knowledge, this is redundant. If not, we should use relative paths.

```suggestion
The result is a new directory `myapp` with a file `src/main.rs` which is the main file for our application.
```

# 3. Adding ICU4X as a dependency

> `ICU4X`'s main meta package is called `icu`, so to start using it, all one has to do it edit their `~/projects/icu_tutorial/myapp/Cargo.toml`. The easiest way to do this is to run:

Reviewer 2: ```suggestion
`ICU4X`'s main meta package is called `icu`, so to start using it, all one has to do it edit their `Cargo.toml`. The easiest way to do this is to run:
```

After saving the changes, commands like `cargo check` will download the `icu` Rust package and use it for the build.

# 4. Locales

`ICU4X` comes with a variety of components allowing to manage various facets of software internationalization.

> Most of those features depend on the selection of a `Locale` which is a particular combination of language, script, region with optional variants. An examples of such locales are `en-US` (American English), `sr-Cyrl` (Serbian with Cyrillic script) or `es-AR` (Argentinian Spanish).

Reviewer 2: The Spanish example doesn't add anything to the English one. Let's show off:

```suggestion
Most of those features depend on the selection of a `Locale` which is a particular combination of language, script, region and other information. An examples of such locales are `en-US` (American English), `sr-Cyrl` (Serbian with Cyrillic script) or `ar-EG-u-nu-latn` (Egyptian Arabic with ASCII numerals).
```

In `ICU4X` `Locale` is a part of the `locid` component. If the user needs just this one feature, they can use `icu_locid` crate as a dependency, but since here we already added a dependency on `icu`, we can refer to it via `icu::locid`.

Let's update our application to use it.

> Open `~/projects/icu_tutorial/myapp/src/main.rs` and edit it to:

Reviewer 2: ```suggestion
Open `src/main.rs` and edit it to:
```

        .expect("Failed to parse locale.");

    if loc.id.language.as_str() == "es" {
        println!("¡Hola!");
    }

    println!("You are using: {}", loc);
}
```
*Notice:* Here, `ICU4X` canonicalized the locales's syntax which uses lowercase letters for the language portion.

After saving it, call `cargo run` in `~/projects/icu_tutorial/myapp` and it should display:

```
¡Hola!
You are using: es-AR
```

Congratulations! `ICU4X` has been used to semantically operate on a locale!

> ## Convenience macro

Reviewer: Nit: This doesn't seem important enough to warrant its own heading. Consider just mixing it in as an example in the above section.

    println!("You are using: {}", loc);
}
```

> *Notice:* Here, `ICU4X` canonicalized the locales's syntax which uses lowercase letters for the language portion.

Reviewer 2: Referring to the above^

"¡Hola amigo!" is a really bad string to use, as it does not take the user's gender into account. We're trying to do good i18n.

> After saving it, call `cargo run` in `~/projects/icu_tutorial/myapp` and it should display:

Reviewer 2: ```suggestion
After saving it, call `cargo run` and it should display:
```

```
¡Hola amigo!
You are using: es-AR
```

Congratulations! `ICU4X` has been used to semantically operate on a locale!

## Convenience macro

> The scenario of working with statically declared Locales is common.

Reviewer 2: Either use lowercase "locale", or capitalized code font "`Locale`".

}
```

In this case, the parsing is performed at compilation time, so we don't need to handle an error case. Try passing an malformed identifier, like "foo-bar" and call `cargo check`.

> *Notice:* `locale!` does not support variants or extension tags (e.g. `u-nu-latn`), as storing these requires allocation. If you have such a tag you need to use runtime parsing.

Reviewer: Issue: This statement is outdated. I suggest removing it. The restrictions on what you can and can't put in `locale!` are a detail that is better discovered from docs.rs.

The way `ICU4X` plugs into that dataset is one of its novelties, aiming at making the data management more flexible and enabling better integration in asynchronous environments.

As a result, compared to most internationalization solutions, working with data in `ICU4X` is a bit more explicit. `ICU4X` provides a trait called `DataProvider` (as well as `BufferProvider` and `AnyProvider`) and a number of concrete APIs that implement these traits for different scenarios.
Users are also free to design their own providers that best fit into their ecosystem requirements.

`BufferProvider` and `AnyProvider` abstract over different ways the data may be loaded: `BufferProvider` abstracts over data providers that deserialize data, whereas `AnyProvider` abstracts over data providers that can provide concrete Rust objects, like data providers that use [`databake`](https://docs.rs/databake).

In this tutorial we are going to use ICU4X's test data provider and then move on to a synchronous file-system data provider which uses ICU4X format JSON resource files.

## Test data

> ICU4X's repository comes with pre-generated test data that covers all of its keys for a select set of locales. For production use it is recommended one use the steps in [Generating Data](#generating-data) to generate custom data, but for the purposes of learning how to use ICU4X, it is sufficient to use the data providers exported by `icu_testdata`. `icu_testdata::any()` will produce an `AnyProvider`, and `icu_testdata::buffer()` will produce a `BufferProvider`. The `any()` provider requires fewer features be enabled so we're using that.

Reviewer: Issue: why are you using the `any()` provider in these examples? The use cases for AnyProvider are fairly limited _in Rust_. Rust users should generally be using either `unstable()` or `buffer()`, unless they fall into one of the narrow exceptions in the constructors documentation page.

> In this case, the parsing is performed at compilation time, so we don't need to handle an error case. Try passing an malformed identifier, like "foo-bar" and try to call `cargo check`.

Reviewer 2: ```suggestion
In this case, the parsing is performed at compilation time, so we don't need to handle an error case. Try passing an malformed identifier, like "foo-bar" and call `cargo check`.
```

> *Notice:* `locale!` does not support variants or extension tags, as storing these requires allocation. If you have such a tag you need to use runtime parsing.

Reviewer 2: What are variants? What are extension tags? Give examples

Next, let's add some more complex functionality.

# 5. Basic Data Management

While the locale API is purely algorithmic, many internationalization APIs require more complex data to work. The most common data set used in Unicode Internationalization is called `CLDR` - `Common Locale Data Repository`.

Data management is a complex and non-trivial area which often requires customizations for particular environments and integrations into projects ecosystem.

The way `ICU4X` plugs into that dataset is one of its novelties, aiming at making the data management more flexible and enabling better integration in asynchronous environments.

> In result, compared to most internationalization solutions, working with data in `ICU4X` is a bit more explicit. `ICU4X` provides a trait called `DataProvider` (as well as `BufferProvider` and `AnyProvider`) and a number of concrete APIs that implement these traits for different scenarios.

Reviewer 2: ```suggestion
As a result, compared to most internationalization solutions, working with data in `ICU4X` is a bit more explicit. `ICU4X` provides a trait called `DataProvider` (as well as `BufferProvider` and `AnyProvider`) and a number of concrete APIs that implement these traits for different scenarios.
```

Users are also free to design their own providers that best fit into their ecosystem requirements.

`BufferProvider` and `AnyProvider` abstract over different ways the data may be loaded: `BufferProvider` abstracts over data providers that deserialize data, whereas `AnyProvider` abstracts over data providers that can provide concrete Rust objects, like data providers that use [`databake`](https://docs.rs/databake).

> In this tutorial we are going to use ICU4X's "test" data provider and then move on to a synchronous file-system data provider which uses ICU4X format JSON resource files.

Reviewer 2: "test" is not the name of the provider or crate, so don't quote it.

```suggestion
In this tutorial we are going to use ICU4X's test data provider and then move on to a synchronous file-system data provider which uses ICU4X format JSON resource files.
```

## Test data

> ICU4X's repository comes with pre-generated test data that covers all of its keys for a select set of locales. For production use it is recommended one use the steps in [Generating Data](#generating-data) to generate custom data, but for the purposes of trying stuff out, it is sufficient to use the data providers exported by `icu_testdata`. `icu_testdata::any()` will produce an `AnyProvider`, and `icu_testdata::buffer()` will produce a `BufferProvider`.

Reviewer 3: `optional, n-b`: `but for the purpose of learning how to use ICU4X...`

# 6. Using an ICU4X component

> We're going to try writing an app that uses the `icu::datetime` component to format a date and time in a Japanese locale.

Reviewer 3: `optional, n-b`: `try` doesn't instill confidence here. `We're going to write an application that`" ...

> First, we need to register our choice of the provider in `~/projects/icu_tutorial/myapp/Cargo.toml`, which can be done by running:

Reviewer 2: ```suggestion
First, we need to register our choice of the provider in `Cargo.toml`, which can be done by running:

```sh
$ cargo add icu_testdata
```

which should generate:

```
[dependencies]
icu = "1.0"
icu_testdata = "1.0"
```

We can then use it in our code:

```rust
fn main() {

>     let _provider = icu_testdata::any();

Reviewer 2: This is weird, we're replacing all the "progress" we've made. Can we not iterate on the code we already have, i.e. keeping the locale declaration and print statement, just adding a date to it?

Author: I think the examples get bigger if we do that, and it's really useful to be able to _focus_ on the important part.

Author: I think that's fair, i also think that might need more holistic work

But I do think it's important to show snippets _before_ iterating. The snippets can be more clearly marked as such though.

use icu::datetime::{DateTimeFormatter, options::length};

fn main() {
    let options = length::Bag::from_date_time_style(length::Date::Long, length::Time::Medium);

    let dtf = DateTimeFormatter::try_new_with_any_provider(&icu_testdata::any(), &locale!("ja").into(), options.into())
        .expect("Failed to initialize DateTimeFormatter");

>     let date = DateTime::try_new_gregorian_datetime(2020, 10, 14, 13, 21, 28)

Reviewer: Big Issue: Never use `try_new_gregorian_datetime` and then pass the result to a `DateTimeFormatter` generated from a variable locale. I continue to fear that this is going to be one of the biggest footguns in our ICU4X API decisions. It definitely doesn't belong in our tutorial!

        .expect("Failed to create a datetime.");

>     // DateTimeFormatter works with data from any calendar, we need to cast to DateTime<AnyCalendar>

Reviewer: Issue: This comment is out of date.

Here's an internationalized date!

> *Notice:* By default, `cargo run` builds and runs a `debug` mode of the binary. If you want to evaluate performance, memory or size of this example, use `cargo run --release`. Our example is also using resources in the `json`  format. It is recommended you generate the data in the `postcard` format (and use it with `BlobDataProvider`) for better performance.

Reviewer 2: No it's not, it's using baked data. I also think it's too early to suggest data formats, the reader doesn't know what those are yet.

# 7. Advanced Data Management

## Using data from the filesystem

If you have ICU4X data on the file system in a JSON format, it can be loaded via `FsDataProvider`, which needs the `icu_provider_fs` crate.

This provider performs deserialization so is a `BufferProvider`. This means that the feature `"serde"` needs to be enabled on `icu` (or the specific `icu_foo` component crate), and `"deserialize_json"` needs to be enabled on `icu_provider`. There are also `"deserialize_postcard_1"` and `"deserialize_bincode_1"` features available.

The whole thing can be done by running:

```sh
$ cargo add icu --features serde
$ cargo add icu_provider_fs
$ cargo add icu_provider --features deserialize_json
```

You may also run `cargo remove icu_testdata` as we don't need it anymore.

This should generate:

```toml
[dependencies]
icu = { version = "1.0", features = ["serde"] }
icu_provider_fs = { version = "1.0.0" }
icu_provider = { version = "1.0.0" , features = ["deserialize_json"] }
```

We can then use it like so:

```rs
use icu_provider_fs::FsDataProvider;

> 

Reviewer 2: This snippet is not runnable, maybe it's too early to show it? We should talk about data generation first

Author: It is, we mention having an icu4x clone

Reviewer 2: *If* you have ICU4X data on the file system in a JSON format, it can be loaded via `FsDataProvider`.

I still think `FsDataProvider` should be the last provider anybody uses, so it should not be our first provider example.

Production users are also recommended to use `BlobDataProvider` from `icu_provider_blob`, which allows a binary blob of (usually `postcard` format) data to be loaded from memory. This data provider provides the flexibility of controlling where the data is stored; allowing for data to even be loaded lazily over the network.

Here is a full example:

```rust
use icu::locid::locale;
use icu::calendar::DateTime;
use icu::datetime::{DateTimeFormatter, options::length};
use icu_provider_fs::FsDataProvider;

fn main() {
    let provider = FsDataProvider::try_new("/path/to/data")
       .expect("Failed to initialize Data Provider.");

    let options = length::Bag::from_date_time_style(length::Date::Long, length::Time::Medium);

    let dtf = DateTimeFormatter::try_new_with_buffer_provider(&provider, &locale!("ja").into(), options.into())
        .expect("Failed to initialize DateTimeFormatter");

>     let date = DateTime::try_new_gregorian_datetime(2020, 10, 14, 13, 21, 28)

Reviewer: Issue: Don't use `try_new_gregorian_datetime`

}
```

You will also need to use the `with_buffer_provider` constructors, e.g. `DateTimeFormatter::try_new_with_buffer_provider()`.

> The ICU4X repository has test data checked in tree in `provider/testdata/data`, however it is recommended one generate data on their own as described in the [next section](#generating-data). Under the hood, `icu_testdata` is simply loading this data.

Reviewer 2: "checked in tree" isn't relevant to the user because they have not checked out our tree.
```suggestion
```

Author: I changed it to mention checking it out

> Production users are recommended to use `BlobDataProvider` from `icu_provider_blob`, which allows a binary blob of (usually `postcard` format) data to be loaded from memory. This data provider provides the flexibility of controlling where the data is stored; allowing for data to even be loaded lazily over the network.

Reviewer 2: Why introduce JSON first if it's not recommended?

Author: Easier to read

Here is a full example:

```rust
use icu::locid::locale;
use icu::calendar::DateTime;
use icu::datetime::{DateTimeFormatter, options::length};
use icu_provider_fs::FsDataProvider;

fn main() {
    let date = DateTime::try_new_gregorian_datetime(2020, 10, 14, 13, 21, 28)
        .expect("Failed to create a datetime.");

    let options = length::Bag::from_date_time_style(length::Date::Long, length::Time::Medium);

>     let provider = FsDataProvider::try_new("/path/to/data")

Reviewer 2: I don't think we should use `FsDataProvider` in examples, it's our worst provider.

Reviewer: Also, the heading above this example recommends using BlobDataProvider, but then the example uses FsDataProvider.

       .expect("Failed to initialize Data Provider.");

    let date = date.to_any();

    let dtf = DateTimeFormatter::try_new_with_buffer_provider(&provider, &locale!("ja").into(), options.into())
        .expect("Failed to initialize DateTimeFormatter");

>     let formatted_date = dtf.format(&date).expect("Formatting should succeed");

Reviewer 3: `optional, n-b`: first initialize `dtf`, then `date` - my brain was truing to understand how the constructor uses `date`.

cargo run --bin icu4x-datagen --features bin -- \
    --cldr-tag latest \
    --icuexport-tag latest \

>     --out ~/projects/icu_tutorial/icu4x-data \

Reviewer 2: ```suggestion
    --out data \
```

Reviewer 2: Nope

Author: Not from a technical POV, I mean from a tutorial POV, i think this absolute path is fine since it's not in the `myapp` folder

It's a reminder that we're stepping out of that folder.

* Then we pass `--cldr-tag` which informs the program which CLDR version to use
* Then we pass `--icuexport-tag` which informs the program which ICU-exported data version to use
* Then we pass `--out` directory which is where we want the generated ICU4X data to be stored

> * Finally, we set `--all-keys --all-locales` which specify that we want to export all keys available, for locales.

Reviewer 2: ```suggestion
* Finally, we set `--all-keys --all-locales` which specify that we want to export all keys available, for all locales.
```

> After that step, it should be possible to navigate to `~/projects/icu_tutorial/icu4x-data` and there should be a `manifest.json` file, and directories with data.

Reviewer 2: ```suggestion
After that step, it should be possible to navigate to `data`, which will contain a `manifest.json` file and a tree of JSON files.
